src/middlewares/cache.ts

```diff
@@ -0,0 +1,92 @@
+import type { MiddlewareFn } from "./mod.ts";
+import type { FreshContext } from "../mod.ts";
+
+/**
+ * Options for {@linkcode cache}.
+ */
+export interface CacheOptions<State> {
+  /**
+   * Whether the cache should be used for the request.
+   *
+   * @param ctx The context of the request.
+   * @returns A boolean indicating whether the cache should be used.
+   *
+   * @example Basic usage
+   * ```ts
+   * import { App, cache } from "fresh";
+   *
+   * const app = new App()
+   *  .use(cache(await caches.open("my-cache"), {
+   *    include: (ctx) => ctx.url.pathname === "/cacheable",
+   *  }))
+   *  .get("/cacheable", () => new Response("Cached!"))
+   *  .get("/uncacheable", () => new Response("Not cached!"));
+   *
+   * await app.listen();
+   * ```
+   */
+  include?: (ctx: FreshContext<State>) => boolean;
+}
+
+/**
+ * Middleware to cache `GET` requests, based on the
+ * {@link https://developer.mozilla.org/en-US/docs/Web/API/Cache | Cache API}.
+ * The underlying caching mechanism does not yet respect the
+ * {@linkcode https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Vary | Vary}
+ * header and takes search parameters into account.
+ *
+ * The {@linkcode CacheOptions.include} option can be used to determine whether
+ * the current request should be cached or not based on the context argument.
+ *
+ * @param cache The cache to use.
+ * @param options Options for the cache middleware.
+ * @returns A middleware function that caches `GET` requests.
+ *
+ * @example Cache the whole site
+ *
+ * Here, `GET` requests to `/` will return the same response.
```

Comment:
The example title and the description here kinda conflicts a bit. I think it can be less confusing if the description here is changed to something like:

```suggestion
 * @example Cache the whole site
 *
 * All `GET` requests will be cached, including File-System routes.
 * Thus requests to `/` will return the same response as long as
 * a cache entry for the `Request` exists.
```

It's also not explicit in the existing JSDoc that the cache is best-effort,
and is purely a performance optimization. Not a guarantee to always return
the same data. 🤔

Reply:
Great suggestions. Thanks!